New cf.read kwarg enabling CDL string input

sadielbartholomew opened the pull request:
Close #171 using the (same) new keyword as suggested in the opening comment of that tagged Issue and in doing so, also improve the error handling for invalid CDL inputs generally.

Note that the API update with the new `cdl_string` keyword argument is (admittedly) a little clunky, given the core (and only non-keyword) argument is named `files` and always accepts file locations other than with this new keyword set to `True` where it then changes to instead accept a string of CDL input (though strictly OPenDAP URLs are also possible to already break the general pattern too, so maybe it is not so bad). However, the only alternative in practice would be to create a whole new function, `cf.read_from_string` or similar, for the case of reading from a string, only possible with CDL anyway.

So, the approach implemented in this PR to enable the feature of #171 is the most reasonable, and certainly "good enough", I think?

sadielbartholomew commented:
@davidhassell as you are aware I am largely working on ES-DOC at the moment but I have been addressing some small "low-hanging fruit" items from the issue trackers on the CF libraries, such as this, as a break here and there.

This one is simple enough in implementation but I thought it would be good to check you are happy with the approach of the new kwarg. I am happy to change the approach if you have a better idea to implement the feature of #171!

davidhassell commented on cf/read_write/read.py:
```diff
@@ -207,6 +210,10 @@ def read(
 
+            As a special case, if the `cdl_string` parameter is set to
+            True, the interpretation of `files` changes so that it
+            should be a string of CDL input rather than the above.
```
To accommodate multiple strings, how about:
```
            As a special case, if the `cdl_string` parameter is set to
            True, the interpretation of `files` changes so that each value
            is assumed to be a string of CDL input rather than the above.
```

davidhassell reviewed (commented):
Hi Sadie,

Thanks - I think the approach is sound - I have made a few comments on the structure.

it would be good to copy behaviour to cfdm, too, which we might want to do in parallel, e.h. because some of the error checking can be moved to there.

sadielbartholomew commented on cf/read_write/read.py:
```diff
@@ -610,6 +641,17 @@ def read(
     field_counter = -1
     file_counter = 0
 
+    if cdl_string:
```
Very good idea, thanks David. I'll get this (or something that does the equivalent) implemented.

Just to be clear, when you say "sequence" you mean as in a Python sequence of CDL strings? In contrast to some string that captures the CDL for a sequence of mutliple netCDF files, if that is even possible? I.e. something like this as file content and/or a string, I am not sure if it is sensible and/or parsable by common tools:

```js
  netcdf example {   // CDL of file 1
...<content>...
  }
  netcdf example {   // CDL of file 2
...<content>...
  }
```

I'll investigate your code snippet (and/or processing a hybrid CDL as above) and for now implement whatever seems in practice to be the intention, but if you have a chance a confirmation as to which would be good. Thanks.

sadielbartholomew replied:
Actually, sorry, ignore my question above, your comment https://github.com/NCAS-CMS/cf-python/pull/227#discussion_r666051349 makes it clear what  was meant, as I will add in a new commit.

davidhassell replied:
> Just to be clear, when you say "sequence" you mean as in a Python sequence of CDL strings?

Yes
> In contrast to some string that captures the CDL for a sequence of mutliple netCDF files, if that is even possible?

I don't know if that's possible! I'm guessing not.

sadielbartholomew commented on cf/read_write/read.py:
```diff
@@ -610,6 +641,17 @@ def read(
+        # The validity of the CDL is tested later with attempt to
+        # convert it to netCDF.
+        c = tempfile.NamedTemporaryFile(
+            mode="wb", dir=tempfile.gettempdir(), prefix="cfdm_", suffix=".cdl"
```
Oops, good spot - that was a mistake left in from a copy and paste. `'w'` and `cf_` are indeed the right parameters.

sadielbartholomew commented on cf/read_write/read.py:
```diff
@@ -930,7 +972,20 @@ def _read_a_file(
         ftype = "netCDF"
         cdl_filename = filename
-        filename = netcdf.cdl_to_netcdf(filename)
+        try:
```
Hmm, I am not entirely sure what you mean by 'new block', sorry - is it that the `try/except` is unnecessary? I'll check to see that. If you meant something different please let me know. Thanks.

davidhassell replied:
Ah, sorry - ignore that comment - I had confused myself!

sadielbartholomew commented on cf/read_write/read.py:
```diff
@@ -930,7 +972,20 @@ def _read_a_file(
+        try:
+            filename = netcdf.cdl_to_netcdf(filename)
+        except subprocess.CalledProcessError as error:
+            msg = str(error)
```
Good idea. I actually didn't realise that function was in `cfdm` as I didn't think `cfdm` handled CDL, not having a `fmt` keyword argument. I'll move it to `cfdm`.

sadielbartholomew replied:
Good idea. I was trying to avoid touching `cfdm` but it seems like a good use case.

sadielbartholomew replied:
Good point, thanks, I will move it.

sadielbartholomew replied:
Moved to cfdm via a new PR: NCAS-CMS/cfdm#147.

sadielbartholomew commented:
I did not realise we were waiting on a merge conflict here until now :grimacing:. I have just resolved it.

All feedback should have been addressed, and all tests are passing locally, so I am now opening and closing the PR to trigger the CI jobs...

sadielbartholomew commented:
Many of the jobs appeared to be hanging (taking much longer than plausible to complete), so I have cancelled them to save on Actions resource. I will investigate the hanging later. As the one job running the test suite to complete passed, along with the linting job, on top of local passes all-round, I am satisfied the PR is sound.

Thanks for your careful review @davidhassell. I believe I have addressed all of your feedback. Since you are on leave and won't see this until next week at the earliest, I will assume all is good and merge this, but any further or retrospective feedback or issues emerging can be addressed in future commits.
